### PR TITLE
Added approval step after deploying to each stage

### LINKED ISSUE #340: Add the capability for customers to run integ test in archer pipeline
Make it possible for customers to write an integ test buildspec and run it in the CodePipeline

### PR DESCRIPTION
**Pull request by hencrice**
Enable customers to run their buildspec_integ.yml after deploying to each stage. The CodeBuild project is provided with the name of the stage that it is validating against. 

Closes #340.

Screen Shot 2019-11-13 at 11 29 36 AM

Screen Shot 2019-11-13 at 11 35 47 AM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by SoManyHs** [comment]
For my own edification, Is this what code pipeline recommends as best practices for running integ tests per stage?

**Comment by hencrice** [comment]
> For my own edification, Is this what code pipeline recommends as best practices for running integ tests per stage?

Hmm, I not aware of any best practice since it's largely based on the needs. The only reason I did it this way is because we assume:
1. all apps in the same stage can be deployed concurrently & independently
2. multiple apps may need to communicate with each other within an Environment to perform business logic.

And that's why I decided to validate the entire stage as a whole.

**Comment by kohidave** [comment]
Actually, could we include the build project in the same stage?

If we do that, it'll stop new changes from coming in and potentially interfering with the tests.

**Comment by hencrice** [comment]
> Actually, could we include the build project in the same stage?
> 
> If we do that, it'll stop new changes from coming in and potentially interfering with the tests.

Include a build action per deployment stage?

File: templates/cicd/pipeline_cfn.yml [hunk 1]
```diff
@@ -93,10 +93,46 @@ Resources:
...
+        Statement:
+          -
+            Effect: Allow
+            Principal:
+              Service:
+                - codebuild.amazonaws.com
+            Action:
+              - sts:AssumeRole
+      Path: /{{$length := len .Stages}}{{if gt $length 0}}{{range $stage := .Stages}}
+  IntegTestProject{{$stage.Name}}:
```
Comment (efekarakus) [inline templates/cicd/pipeline_cfn.yml]: Is it possible that the stage name contains a dash? if so we might want to do something similar to:
https://github.com/aws/amazon-ecs-cli-v2/blob/e19329b06fcf1c77d048fa7d85f803d0fa3bcb4f/internal/pkg/deploy/cloudformation/stack/template_functions.go#L12-L22

Comment (hencrice) [reply templates/cicd/pipeline_cfn.yml]: lol sure :)

Comment (hencrice) [reply templates/cicd/pipeline_cfn.yml]: Oh actually the stage name can contain dash. Take a look at the screenshot in the "conversation" tab

**Comment by kohidave** [comment]
We should have a follow up discussion on where exactly these buildspecs are defined - but this is a good first go!

File: templates/cicd/pipeline_cfn.yml [hunk 2]
```diff
@@ -93,10 +93,46 @@ Resources:
...
+        ComputeType: BUILD_GENERAL1_SMALL
+        PrivilegedMode: true
+        Image: aws/codebuild/amazonlinux2-x86_64-standard:1.0
+        EnvironmentVariables:
+          - Name: PIPELINE_STAGE
+            Type: PLAINTEXT
+            Value: {{$stage.Name}}
+      Source:
+        Type: CODEPIPELINE
+        BuildSpec: ecs-project/buildspec_integ.yml
```
[the remark this answers is not in the document]
Comment (hencrice) [reply templates/cicd/pipeline_cfn.yml]: nope the action will fail in that case

**Pull request merged**